### PR TITLE
Add unit test for duplicate message ID checks

### PR DESCRIPTION
**Pull request by rrealmuto**
Because we keep running into this problem.

Adds a runtime check that is capable of detecting duplicate messages being updated/added. Generation will throw an exception if a duplicate is detected:

Screenshot 2023-05-20 130331

Also adds a more basic CI check that just checks the static tables.

### CODE REVIEW

**Comment by fenhl** [comment]
I'd suggest moving this into CI since it's a fast check that doesn't require rolling a seed. It also doesn't need to be called from Patches.

**Comment by rrealmuto** [comment]
I included the call in Patches for development purposes (that's where I originally put the check so I would see every time it broke). Maybe we keep it there in case someone decides to implement custom message plando ;)

As for moving it into CI, I'm not sure I follow. Isn't that exactly what I did by adding a unit test? Or do you mean add it to the --no-unit-tests part of CI?

**Comment by fenhl** [comment]
> I included the call in Patches for development purposes (that's where I originally put the check so I would see every time it broke). Maybe we keep it there in case someone decides to implement custom message plando ;)

I don't follow. How would that help in that case? And even then, I think it's unnecessary right now (at least with my other suggested change), since it's re-checking something that's already handled by CI. All it does is add a small amount of time to seed generation.

> As for moving it into CI, I'm not sure I follow. Isn't that exactly what I did by adding a unit test? Or do you mean add it to the --no-unit-tests part of CI?

Yes, that's what I mean. CI with `--no_unit_tests` is a lot faster than running the entire thing so it's something that people can add as pre-commit hooks.

**Comment by rrealmuto** [comment]
So I thought about your comment on how it was still missing messages that are modified from within Patches.py. I updated the runtime check to account for that so I think now it's definitely not redundant. If anything, now the CI one is 😂

**Comment by mracsys** [comment]
A few more places that this trips in Patches.py that should have `allow_duplicates=True`:
1. Maps/Compasses give info enabled, there are three calls in the if block starting on [line 2271](https://github.com/OoTRandomizer/OoT-Randomizer/blob/399123e6c7d3877083886d80ebf7e6e81b542a8a/Patches.py#L2271)
2. Blue Fire Arrows enabled, one call on [line 2396](https://github.com/OoTRandomizer/OoT-Randomizer/blob/399123e6c7d3877083886d80ebf7e6e81b542a8a/Patches.py#L2396)
3. (For documentation, not relevant to the main branch) On Roman's branch, the fast token changes on [line 2360](https://github.com/Roman971/OoT-Randomizer/blob/284d2df92d4d980c55fa50dca43b2b52f7956327/Patches.py#L2360)

**Comment by cjohnson57** [comment]
This is definitely a great check considering how often this issue trips us up, lol

I agree that it's better to have the non-CI check so that you can catch it before opening a PR. 


> A few more places that this trips in Patches.py that should have `allow_duplicates=True`:
> 
> 1. Maps/Compasses give info enabled, there are three calls in the if block starting on [line 2271](https://github.com/OoTRandomizer/OoT-Randomizer/blob/399123e6c7d3877083886d80ebf7e6e81b542a8a/Patches.py#L2271)
> 2. Blue Fire Arrows enabled, one call on [line 2396](https://github.com/OoTRandomizer/OoT-Randomizer/blob/399123e6c7d3877083886d80ebf7e6e81b542a8a/Patches.py#L2396)
> 3. (For documentation, not relevant to the main branch) On Roman's branch, the fast token changes on [line 2360](https://github.com/Roman971/OoT-Randomizer/blob/284d2df92d4d980c55fa50dca43b2b52f7956327/Patches.py#L2360)

So the feature as-is gives false positives with these settings?

**Comment by mracsys** [comment]
It does for the first two settings in this branch if they are turned on. I have a PR to fix them here: https://github.com/rrealmuto/OoT-Randomizer/pull/78

**Review by cjohnson57 — APPROVED** [review]

File: CI.py [hunk 1]
```diff
@@ -79,6 +80,22 @@ def check_hell_mode_tricks(fix_errors=False):
...
+# For that, we perform additional checking at patch time
+def check_message_duplicates():
+    def check_for_duplicates(new_item_messages):
+        for i in range(0, len(new_item_messages)):
+            for j in range(i, len(new_item_messages)):
+                if i != j:
+                    message_id1, message1 = new_item_messages[i]
+                    message_id2, message2 = new_item_messages[j]
+                    if message_id1 == message_id2:
+                        error("Duplicate MessageID found: " + hex(message_id1) + ", " + message1 + ", " + message2, False)
```
Comment (fenhl) [inline CI.py:94]:
```suggestion
                        error(f'Duplicate MessageID found: {hex(message_id1)}, {message1}, {message2}', False)
```

File: Messages.py [hunk 2]
```diff
@@ -861,7 +862,14 @@ def from_bytearray(cls, bytearray, id=0, opts=0x00):
 
 # wrapper for updating the text of a message, given its message id
 # if the id does not exist in the list, then it will add it
-def update_message_by_id(messages, id, text, opts=None):
+# Checks if the message being updated is a newly added message in order to prevent duplicates.
+# Use allow_duplicates=True if the same message is purposely updated multiple times
+def update_message_by_id(messages, id, text, opts=None, allow_duplicates=False):
+    # Check is we have previously added/modified this message.
+    if id in new_messages and not allow_duplicates:
+        raise Exception("Attempting to add duplicate message " + hex(id))
```
Comment (fenhl) [inline Messages.py:870]:
```suggestion
        raise Exception(f'Attempting to add duplicate message{hex(id)}')
```

File: Patches.py [hunk 3]
```diff
@@ -2162,7 +2163,7 @@ def update_scrub_text(message, text_replacement, default_price, price, item_name
             location = world.get_location("Market Treasure Chest Game Salesman")
             item_text = getHint(getItemGenericName(location.item), True).text
             update_message_by_id(messages, 0x6D, "I seem to have misplaced my\x01keys, but I have a fun item to\x01sell instead.\x04How about \x05\x4110 Rupees\x05\x40 for\x01\x05\x41" + item_text + "\x05\x40?\x01\x1B\x05\x42Buy\x01Don't Buy\x05\x40\x02")
-        update_message_by_id(messages, 0x908B, "That's OK!\x01More fun for me.\x0B\x02", 0x00)
+        update_message_by_id(messages, 0x908B, "That's OK!\x01More fun for me.\x0B\x02", 0x00, True)
```
Comment (fenhl) [inline Patches.py:2166]:
```suggestion
        update_message_by_id(messages, 0x908B, "That's OK!\x01More fun for me.\x0B\x02", 0x00, allow_duplicates=True)
```

**Review by fenhl — CHANGES_REQUESTED** [review]
A couple code style changes, looks good other than that.
